spanconfig: introduce spanconfig.Reporter

irfansharif opened the pull request:
This is KV-side API for multi-tenant replication reports (#89987).


```go
 // Reporter generates a conformance report over the given spans, i.e. whether
 // the backing ranges conform to the span configs that apply to them.
 //
 // NB: The standard implementation does not use a point-in-time snapshot of span
 // config state, but could be made to do so if needed. See commentary on the
 // spanconfigreporter.Reporter type for more details ("... we might not have a
 // point-in-time snapshot ...").
 type Reporter interface {
 	SpanConfigConformance(
 		ctx context.Context, spans []roachpb.Span,
 	) (roachpb.SpanConfigConformanceReport, error)
 }
```

```proto
// SpanConfigConformanceReport reports ranges that (i) don't conform to span
// configs that apply over them, and (ii) are unavailable. Also included in this
// report are the IDs of unavailable nodes (possibly contributing to
// under-replication or range-unavailability).
message SpanConfigConformanceReport {
  repeated ConformanceReportedRange under_replicated = 1 [(gogoproto.nullable) = false];
  repeated ConformanceReportedRange over_replicated = 2 [(gogoproto.nullable) = false];
  repeated ConformanceReportedRange violating_constraints = 3 [(gogoproto.nullable) = false];
  repeated ConformanceReportedRange unavailable = 4 [(gogoproto.nullable) = false];
  repeated int32 unavailable_node_ids = 5 [(gogoproto.customname) = "UnavailableNodeIDs"];
};

message ConformanceReportedRange {
  RangeDescriptor range_descriptor = 1 [(gogoproto.nullable) = false];
  SpanConfig config = 2 [(gogoproto.nullable) = false];
}
```

irfansharif committed:
spanconfig: introduce spanconfig.Reporter

This is KV-side API for multi-tenant replication reports (#89987)

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/90016)
<!-- Reviewable:end -->

irfansharif commented:
@arulajmani, this is still incomplete and slightly buggy, but it should be ready for a look. The incompleteness is marked explicitly (`XXX:`) and I'll address as part of this PR.

arulajmani reviewed (commented):
Good stuff! Should we add authorization for the new RPCs in `auth_tenant.go` as part of this patch? 

Separately, I remember there was some talk about pulling all the information to compute conformance into the tenant and constructing the report there as opposed to having KV perform the computation like this patch does -- could you remind me why we settled on this approach? 

> Reviewed 40 of 42 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-:-NFATw31DA_xDQB7GJiC:b484pnh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif)

___
*[`pkg/kv/kvserver/store.go` line 1076 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NF9urQkABh2WwxZ3DIJ:-NF9urQkABh2WwxZ3DIK:bys3ki2) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c1de07e122ac8284febed7d183494a56219cdd4/pkg/kv/kvserver/store.go#L1076)):*
> ```Go
> 	// infrastructure or not.
> 	//
> 	// TODO(irfansharif): We can remove this.
> ```

Let's file an issue for this? 
___
*[`pkg/kv/kvserver/reports/replication_stats_report.go` line 407 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NF9v7lT7MKxt5JPZgnQ:-NF9v7lT7MKxt5JPZgnR:b6wuc0j) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c1de07e122ac8284febed7d183494a56219cdd4/pkg/kv/kvserver/reports/replication_stats_report.go#L407)):*
> ```Go
> 		// on non-voting replicas. This code will also soon be removed in favor
> 		// of something that works with multi-tenancy (#89987).
> 	}, replicationFactor, 0)
> ```

nit: /*neededNonVoters*/
___
*[`pkg/roachpb/metadata_replicas.go` line 415 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NFAPz-O4yh0NoEqXxih:-NFAPz-O4yh0NoEqXxii:b-eclr8w) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c1de07e122ac8284febed7d183494a56219cdd4/pkg/roachpb/metadata_replicas.go#L415)):*
> ```Go
> // over/under-replication of voters. If the caller is only interested in
> // availability of voting replicas, 0 can be passed in. neededNonVoters is the
> // counterpart for non-voting replicas.
> ```

Should we use -1 as the sentinel value for non-voters? (and maybe make the change for voters as well, for parity). 

Unlike voters, it's completely reasonable to expect 0 non-voters. Should we be detecting cases where that's the expectation but non-voters exist as `OverReplicatedNonVoters`? 
___
*[`pkg/spanconfig/spanconfig.go` line 356 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NFAnpKS5pPFIU4zR1TC:-NFAnpKS5pPFIU4zR1TD:bfkrr7f) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c1de07e122ac8284febed7d183494a56219cdd4/pkg/spanconfig/spanconfig.go#L356)):*
> ```Go
> // the backing ranges conform to the span configs that apply to them.
> type Reporter interface {
> 	SpanConfigConformance(
> ```

Could you add some more words about the `spans` field? Specifically, I was thinking something that talks about how conformance isn't reported at a "snapshot" for all spans passed into this function and why that's okay.
___
*[`pkg/spanconfig/spanconfigreporter/reporter.go` line 102 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NFAcTyP1U-xAQ0E8DyY:-NFAcTyP1U-xAQ0E8DyZ:b9dv3dy) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c1de07e122ac8284febed7d183494a56219cdd4/pkg/spanconfig/spanconfigreporter/reporter.go#L102)):*
> ```Go
> 	// XXX: Write an end-to-end test using actual SQL and zone configs. Set configs
> 	// on a table, disable replication, see conformance. Enable repl, change
> 	// configs, etc. Use tenants as well for this mode. Do this for tenants as well.
> ```

Would this have caught the `auth_tenant.go` omission? 
___
*[`pkg/spanconfig/spanconfigreporter/reporter.go` line 107 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NFAd9Gh8LkKigHrdLj9:-NFAd9Gh8LkKigHrdLjA:bfq6lys) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c1de07e122ac8284febed7d183494a56219cdd4/pkg/spanconfig/spanconfigreporter/reporter.go#L107)):*
> ```Go
> 	// XXX: Can we improve the SpanConfigConformanceReport proto type? Perhaps
> 	// include some {meta,}data about the span config being violated as well? Or
> 	// include the span config directly and provide helper libraries to compute
> ```

I'm not sure how this API will be used yet, but including the span config seems useful in understanding why things are in violation, especially given the various steps involved in going from a zone config to the range's placement. 
___
*[`pkg/spanconfig/spanconfigreporter/reporter.go` line 109 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NFAnQj1B-hZ49xcj4Nx:-NFAnQj1B-hZ49xcj4Ny:bs6tv4b) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c1de07e122ac8284febed7d183494a56219cdd4/pkg/spanconfig/spanconfigreporter/reporter.go#L109)):*
> ```Go
> 	// include the span config directly and provide helper libraries to compute
> 	// human-readable "why is this in violation" text.
> 	// - Only include range ID + replica descriptors + keys?
> ```

Is this simply to avoid shipping the entire range descriptor back or is there something else?
___
*[`pkg/spanconfig/spanconfigreporter/reporter.go` line 111 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NFA_Iak67RIK6I8n-ar:-NFA_Iak67RIK6I8n-as:bp14uhz) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c1de07e122ac8284febed7d183494a56219cdd4/pkg/spanconfig/spanconfigreporter/reporter.go#L111)):*
> ```Go
> 	// - Only include range ID + replica descriptors + keys?
> 	// - Type to represent exactly which constraint exactly is being violated?
> 	// - Segment over/under replicated by what replica type (voter/non-voter)
> ```

At the least, I think we should do this, especially given we already have this information and we're intentionally coalescing things. Previously if a range appeared as both under and over-replicated we knew exactly why that was the case -- that's no longer true now that this reporting stuff also knows about non-voters and I think the UX here is a bit confusing. You could also argue users care more about voter under/over-replication compared to non-voters, so the distinction is useful to make. What do you think? 
___
*[`pkg/spanconfig/spanconfigreporter/reporter.go` line 155 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NFAizNJCEXEojYJVDxQ:-NFAizNJCEXEojYJVDxR:b-mij4ir) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c1de07e122ac8284febed7d183494a56219cdd4/pkg/spanconfig/spanconfigreporter/reporter.go#L155)):*
> ```Go
> 			for i, c := range overall.Constraints {
> 				if c.NumReplicas == 0 {
> 					c.NumReplicas = conf.NumReplicas
> ```

nit: Could you add a comment to make this logic a bit more obvious please. 

nit2: Do you think it be clearer if we created a new variable instead of assigning `NumReplicas`?
___
*[`pkg/kv/kvclient/kvcoord/send_test.go` line 107 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NFAb_1A3Q9lKdqLo0AX:-NFAb_1A3Q9lKdqLo0AY:b-smzzv4) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c1de07e122ac8284febed7d183494a56219cdd4/pkg/kv/kvclient/kvcoord/send_test.go#L107)):*
> ```Go
> 	context.Context, *roachpb.SpanConfigConformanceRequest,
> ) (*roachpb.SpanConfigConformanceResponse, error) {
> 	panic("implement me")
> ```

Should this be s/implement me/unimplemented/g or were you needing to do this as part of this patch?
___
*[`pkg/kv/kvserver/allocator/allocatorimpl/allocator_test.go` line 3970 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NFAbBXd3z3vY-wyTdGH:-NFAbBXd3z3vY-wyTdGI:b9t4oz3) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c1de07e122ac8284febed7d183494a56219cdd4/pkg/kv/kvserver/allocator/allocatorimpl/allocator_test.go#L3970)):*
> ```Go
> 			}
> 		}
> 		analyzed := constraint.AnalyzeConstraints(a.StorePool, existingRepls, 0, tc.constraints)
> ```

nit: /* numReplicas */


<!-- Sent from Reviewable.io -->

irfansharif committed:
*: review fixup (will be squashed in)

irfansharif reviewed (commented):
> Should we add authorization for the new RPCs in auth_tenant.go as part of this patch?

Done.

>  I remember there was some talk about pulling all the information to compute conformance into the tenant and constructing the report there as opposed to having KV perform the computation like this patch does -- could you remind me why we settled on this approach?

I wasn't sure how much pre-computing this periodically and stashing the result somewhere was actually buying us. In a future PR/benchmark I want to run this conformance report against a million or so ranges and see how it fares -- I suspect it'll be good enough to not need pre-computing. BTW we can still pre-compute data and store it, the way we'd do it is kick off a job that invokes this lower-level API and stashes the result somewhere in SQL/ObsService. This API here though is still just the lower level thing and does not try to do persistence. I felt it was simpler.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-:-NGJ4uHS2oMhTi2a-vgC:b18dqdv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/kv/kvserver/store.go` line 1076 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NF9urQkABh2WwxZ3DIJ:-NGHTO1EC-7Cu4DGmLyX:blv6syr) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c1de07e122ac8284febed7d183494a56219cdd4/pkg/kv/kvserver/store.go#L1076)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Let's file an issue for this? 
</blockquote></details>

Added to https://github.com/cockroachdb/cockroach/issues/81009.
___
*[`pkg/kv/kvserver/reports/replication_stats_report.go` line 407 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NF9v7lT7MKxt5JPZgnQ:-NGJ5f_a9w663in5-r3g:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c1de07e122ac8284febed7d183494a56219cdd4/pkg/kv/kvserver/reports/replication_stats_report.go#L407)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: /*neededNonVoters*/
</blockquote></details>

Done.
___
*[`pkg/roachpb/metadata_replicas.go` line 415 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NFAPz-O4yh0NoEqXxih:-NGHUNUH87Js2f42YMUu:btv6bbh) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c1de07e122ac8284febed7d183494a56219cdd4/pkg/roachpb/metadata_replicas.go#L415)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should we use -1 as the sentinel value for non-voters? (and maybe make the change for voters as well, for parity). 

Unlike voters, it's completely reasonable to expect 0 non-voters. Should we be detecting cases where that's the expectation but non-voters exist as `OverReplicatedNonVoters`? 
</blockquote></details>

Good point, I'm going to use -1 as the sentinel value for non-voters. I'll keep the voters version as is to not accidentally introduce new bugs (there's a bit of tacitness in how neededVoters == 0 is handled below that I don't want to unentangle).
___
*[`pkg/spanconfig/spanconfig.go` line 356 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NFAnpKS5pPFIU4zR1TC:-NGHVBEJ1T5fmFhVGVK_:b7753r0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c1de07e122ac8284febed7d183494a56219cdd4/pkg/spanconfig/spanconfig.go#L356)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Could you add some more words about the `spans` field? Specifically, I was thinking something that talks about how conformance isn't reported at a "snapshot" for all spans passed into this function and why that's okay.
</blockquote></details>

I'll point to this comment block to who ever gets to use this interface to power the "front-end" of multi-tenant repl reports, so pardon the verbosity. The data dependencies in the implementation are:
- a view of range descriptors (done so transactionally and probably will continue to be so -- so we have a snapshot);
- a view over what span configs apply to what keyspans; backed by the kvsubscriber machinery you may have heard of. Here it's more interesting in that different nodes might be observing data as of a different timestamp, but that's not quite breaking a "snapshot" of the data. It's true that we aren't grabbing a read lock over the kvsubscriber while reading off each span config, but that's something we can do if it becomes a problem;
- the store resolver resolving store IDs to store descriptors. It's true that we're not making consistency claims over this data -- it's possible that we're racing against a store being added while that info is not available through gossip; the fallback behaviour is [well-defined](https://github.com/cockroachdb/cockroach/blob/f783b08109e076840cdc473e9cf04b45bfea9309/pkg/kv/kvserver/constraint/analyzer.go#L66-L69). You might see such inconsistencies if you're issuing this RPC against different nodes with different views over the gossip data, but I don't think it's worth leaking such details into the interface itself. Each RPC is still working over a "snapshot" in time, at the servicing node.
- gossip-backed node liveness; same as point above.
___
*[`pkg/spanconfig/spanconfigreporter/reporter.go` line 102 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NFAcTyP1U-xAQ0E8DyY:-NGHYxHpDyCmt7EK-9EH:b4xz9c8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c1de07e122ac8284febed7d183494a56219cdd4/pkg/spanconfig/spanconfigreporter/reporter.go#L102)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Would this have caught the `auth_tenant.go` omission? 
</blockquote></details>

I've added the `auth_tenant.go` checks, and yes, it would've caught the omission. I'm not adding these tests now since it's not hooked into SQL/an endpoint -- that'll come in some follow up PR.
___
*[`pkg/spanconfig/spanconfigreporter/reporter.go` line 107 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NFAd9Gh8LkKigHrdLj9:-NGJ47Hb96h0NeTDR-8Y:bohqzsr) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c1de07e122ac8284febed7d183494a56219cdd4/pkg/spanconfig/spanconfigreporter/reporter.go#L107)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I'm not sure how this API will be used yet, but including the span config seems useful in understanding why things are in violation, especially given the various steps involved in going from a zone config to the range's placement. 
</blockquote></details>

Agreed, done.
___
*[`pkg/spanconfig/spanconfigreporter/reporter.go` line 109 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NFAnQj1B-hZ49xcj4Nx:-NGHZFC9CX1kKfcZQEwk:b-z43kys) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c1de07e122ac8284febed7d183494a56219cdd4/pkg/spanconfig/spanconfigreporter/reporter.go#L109)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Is this simply to avoid shipping the entire range descriptor back or is there something else?
</blockquote></details>

Nothing else, these were just (sloppy) notes for myself. Shipping the entire descriptor back seems fine -- we seem to need most of it.
___
*[`pkg/spanconfig/spanconfigreporter/reporter.go` line 111 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NFA_Iak67RIK6I8n-ar:-NGH_-k48xuZoKP_LAnR:b-o9u6tx) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c1de07e122ac8284febed7d183494a56219cdd4/pkg/spanconfig/spanconfigreporter/reporter.go#L111)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

At the least, I think we should do this, especially given we already have this information and we're intentionally coalescing things. Previously if a range appeared as both under and over-replicated we knew exactly why that was the case -- that's no longer true now that this reporting stuff also knows about non-voters and I think the UX here is a bit confusing. You could also argue users care more about voter under/over-replication compared to non-voters, so the distinction is useful to make. What do you think? 
</blockquote></details>

I think the distinction is worth making. I'm going to do something different -- in the report proto itself I'll include the list of unavailable node IDs so callers can 
___
*[`pkg/spanconfig/spanconfigreporter/reporter.go` line 155 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NFAizNJCEXEojYJVDxQ:-NGJ4is359hNMogG66IG:b-q8kx0d) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c1de07e122ac8284febed7d183494a56219cdd4/pkg/spanconfig/spanconfigreporter/reporter.go#L155)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: Could you add a comment to make this logic a bit more obvious please. 

nit2: Do you think it be clearer if we created a new variable instead of assigning `NumReplicas`?
</blockquote></details>

Added a comment. I'll keep the assignment as is -- better to have a single place to learn about NumReplicas instead of two.
___
*[`pkg/kv/kvclient/kvcoord/send_test.go` line 107 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NFAb_1A3Q9lKdqLo0AX:-NFBGo9_EL9RpCl14QbA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c1de07e122ac8284febed7d183494a56219cdd4/pkg/kv/kvclient/kvcoord/send_test.go#L107)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should this be s/implement me/unimplemented/g or were you needing to do this as part of this patch?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
This looks pretty close, just a few minor comments. 

> I wasn't sure how much pre-computing this periodically and stashing the result somewhere was actually buying us. In a future PR/benchmark I want to run this conformance report against a million or so ranges and see how it fares -- I suspect it'll be good enough to not need pre-computing. BTW we can still pre-compute data and store it, the way we'd do it is kick off a job that invokes this lower-level API and stashes the result somewhere in SQL/ObsService. This API here though is still just the lower level thing and does not try to do persistence. I felt it was simpler.

I think we're talking about different things. When I said "compute conformance in the tenant" I meant instead of doing this under-replicated, over-replicated, constraint conformance etc. in KV, we could instead ship back all the tenant's range descriptors, span configs, and the view of node liveness and compute conformance inside the SQL pod. I remember this was discussed in one of our eng pods, though I don't remember if we settled one way or another. Do you remember why we ended up choosing this approach? 

To be clear, I'm not asking you to switch it -- just wondering if you considered the alternative.

> Reviewed 58 of 58 files at r2, 42 of 42 files at r3, 14 of 14 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-:-NHXI-koB60G6Lx0n57S:b-yuu6cg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif)

___
*[`pkg/rpc/auth_tenant.go` line 389 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NHXBbx59JsVZaTOSPsB:-NHXBbx59JsVZaTOSPsC:b-i4i0f9) ([raw file](https://github.com/cockroachdb/cockroach/blob/ddb99140906c6f394a1b249a9d527dde64890f4d/pkg/rpc/auth_tenant.go#L389)):*
> ```Go
> 	case *roachpb.SpanConfigTarget_Span:
> 		return validateSpan(tenID, *spanConfigTarget.GetSpan())
> 	case *roachpb.SpanConfigTarget_SystemSpanConfigTarget:
> ```

What's a system span config target? :trollface: 
___
*[`pkg/spanconfig/spanconfig.go` line 356 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NFAnpKS5pPFIU4zR1TC:-NHX8cHnDnXUpCNOgiPb:bwzrdnq) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c1de07e122ac8284febed7d183494a56219cdd4/pkg/spanconfig/spanconfig.go#L356)):*
> backed by the kvsubscriber machinery you may have heard of.

Lol.

> gossip-backed node liveness; same as point above

I might be missing something, but I left a question in-line for you. 

> It's true that we aren't grabbing a read lock over the kvsubscriber while reading off each span config, but that's something we can do if it becomes a problem;

Could we please add this as a comment in code, instead of trying to reference this thread? It doesn't have to be as verbose as your explanation above, but do get this part -- I think it's important. 

To be clear, I'm not asking you to change any of this behavior or saying it's incorrect -- just add some words for it :) 
___
*[`pkg/spanconfig/spanconfigreporter/reporter.go` line 102 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NFAcTyP1U-xAQ0E8DyY:-NHX7Ym5CiosAfD7q9Wc:b-qir3n0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c1de07e122ac8284febed7d183494a56219cdd4/pkg/spanconfig/spanconfigreporter/reporter.go#L102)):*
> I'm not adding these tests now since it's not hooked into SQL/an endpoint -- that'll come in some follow up PR.

Makes sense. However, before we merge, let's add the authentication specific checks to `auth_test.go` though?
___
*[`pkg/spanconfig/spanconfigreporter/reporter.go` line 111 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NFA_Iak67RIK6I8n-ar:-NHXC6td3FN6fAa-0WUO:b-izssdw) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c1de07e122ac8284febed7d183494a56219cdd4/pkg/spanconfig/spanconfigreporter/reporter.go#L111)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

I think the distinction is worth making. I'm going to do something different -- in the report proto itself I'll include the list of unavailable node IDs so callers can 
</blockquote></details>

Sounds good, I'm all for it. 
___
*[`pkg/spanconfig/spanconfigreporter/reporter.go` line 115 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NHXGCmF6oo7KsizWOx6:-NHXGCmF6oo7KsizWOx7:blph8q1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ddb99140906c6f394a1b249a9d527dde64890f4d/pkg/spanconfig/spanconfigreporter/reporter.go#L115)):*
> ```Go
> 					status := desc.Replicas().ReplicationStatus(
> 						func(rDesc roachpb.ReplicaDescriptor) bool {
> 							isLive, err := r.dep.Liveness.IsLive(rDesc.NodeID)
> ```

Are we guaranteed that `IsLive` returns a constant status for a particular node for every call? Is it possible for a node to flap in and out of liveness while we're iterating through range descriptors? 
___
*[`pkg/spanconfig/spanconfigreporter/datadriven_test.go` line 189 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NHXANsC8kbWe89EoisI:-NHXANsC8kbWe89EoisJ:bmi8ppe) ([raw file](https://github.com/cockroachdb/cockroach/blob/ddb99140906c6f394a1b249a9d527dde64890f4d/pkg/spanconfig/spanconfigreporter/datadriven_test.go#L189)):*
> ```Go
> 							buf.WriteString(fmt.Sprintf("%s:\n", tag))
> 						}
> 						buf.WriteString(fmt.Sprintf("  %s applying %s\n", printRangeDesc(r.RangeDescriptor),
> ```

:100: 


<!-- Sent from Reviewable.io -->

irfansharif committed:
liveness: move IsLiveMap type

We'll depend on just this directly in a future commit, without wanting to
depend on the much larger liveness package. Instead of introducing a
standalone package for it, we'll just re-use the thin livenesspb package
instead.

Release note: None

irfansharif reviewed (commented):
Ah. Hm, I'm not sure what the benefit of doing it that way would be. There is one downside -- we'd be shipping a lot of data over the wire when all we're interested in is just the result. Perhaps one benefit is that the CPU time spent processing this request is tied directly to the tenant pod which is appropriately billed, but I'm not sure that's a strong reason (+ we could always do proper attribution server-side).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-:-NHapAQvAh3Wxx7feWdI:b-aawg2l)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/spanconfig/spanconfig.go` line 356 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NFAnpKS5pPFIU4zR1TC:-NHb4E8s2y0B5mUQHGuq:b222rqs) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c1de07e122ac8284febed7d183494a56219cdd4/pkg/spanconfig/spanconfig.go#L356)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

> backed by the kvsubscriber machinery you may have heard of.

Lol.

> gossip-backed node liveness; same as point above

I might be missing something, but I left a question in-line for you. 

> It's true that we aren't grabbing a read lock over the kvsubscriber while reading off each span config, but that's something we can do if it becomes a problem;

Could we please add this as a comment in code, instead of trying to reference this thread? It doesn't have to be as verbose as your explanation above, but do get this part -- I think it's important. 

To be clear, I'm not asking you to change any of this behavior or saying it's incorrect -- just add some words for it :) 
</blockquote></details>

Done. Added the following:

```
		// NB: The data dependencies in the implementation are:
		//
		// i.   point-in-time view over gossip-backed node liveness;
		// ii.  point-in-time view of range descriptors (done transactionally);
		// iii. the store resolver resolving store IDs to store descriptors;
		// iv.  view over what span configs apply to what keyspans;
		//
		// TODO(irfansharif): For (iii) and (iv) we might not have a
		// point-in-time snapshot of the data.
		// - For (iii) it's possible that as we iterate through the set of range
		//   descriptors, a few of which refer to some store S, we're racing
		//   against that newly-added store's info not yet being available
		//   through gossip. This is exceedingly unlikely, but if we see it
		//   happen, we can expose some snapshot of the StoreResolver state like
		//   we have for liveness.
		// - For (iv) too we're not grabbing a read lock over the backing
		//   spanconfig.KVSubscriber while reading off each span config, so it's
		//   possible we generate the report for two range descriptors with span
		//   configs from different points in time. If this too becomes a
		//   problem, we can explicitly generate a snapshot like we do for
		//   liveness.
```
___
*[`pkg/spanconfig/spanconfigreporter/reporter.go` line 102 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NFAcTyP1U-xAQ0E8DyY:-NHamUVP2zTj4LsPopai:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c1de07e122ac8284febed7d183494a56219cdd4/pkg/spanconfig/spanconfigreporter/reporter.go#L102)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

> I'm not adding these tests now since it's not hooked into SQL/an endpoint -- that'll come in some follow up PR.

Makes sense. However, before we merge, let's add the authentication specific checks to `auth_test.go` though?
</blockquote></details>

Done.
___
*[`pkg/spanconfig/spanconfigreporter/reporter.go` line 111 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NFA_Iak67RIK6I8n-ar:-NHajLo50QcBFEh8Fn88:bqkij1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c1de07e122ac8284febed7d183494a56219cdd4/pkg/spanconfig/spanconfigreporter/reporter.go#L111)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Sounds good, I'm all for it. 
</blockquote></details>

Message got clipped off from earlier, but meant to say "unavailable node IDs so callers can use library functions (not written) to determine exactly why the the specific range is unavailable/non-conformant". Done.
___
*[`pkg/spanconfig/spanconfigreporter/reporter.go` line 115 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NHXGCmF6oo7KsizWOx6:-NHamgko6vSGbQ17Q6zJ:b-ojtvbk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ddb99140906c6f394a1b249a9d527dde64890f4d/pkg/spanconfig/spanconfigreporter/reporter.go#L115)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Are we guaranteed that `IsLive` returns a constant status for a particular node for every call? Is it possible for a node to flap in and out of liveness while we're iterating through range descriptors? 
</blockquote></details>

Great catch! Indeed it can, and that's not something we want. Changed interfaces around slightly to explicitly capture a snapshot of liveness state so such flip-flopping is not possible.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (approved):
> Ah. Hm, I'm not sure what the benefit of doing it that way would be.

One thought was that the APIs required to "build the report" would be more generally useful. For example, shipping all range descriptors back to the tenant can be used for both building conformance and enabling `SHOW RANGES`. However, after thinking about this a bit more, I doubt we want to let tenants have a view of the liveness of the underlying cluster. So what you have here seems good to me. 

Anyway, thanks for the iterations. <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> modulo comments. 

> Reviewed 72 of 72 files at r5, 58 of 58 files at r6, 31 of 31 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-:-NHjble-9RU0xY3UAuGT:b-rwi6qm)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @irfansharif)

___
*[`pkg/spanconfig/spanconfig.go` line 356 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NFAnpKS5pPFIU4zR1TC:-NHjYrWzAKK38mj1ZneV:brqvvq1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c1de07e122ac8284febed7d183494a56219cdd4/pkg/spanconfig/spanconfig.go#L356)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Done. Added the following:

```
		// NB: The data dependencies in the implementation are:
		//
		// i.   point-in-time view over gossip-backed node liveness;
		// ii.  point-in-time view of range descriptors (done transactionally);
		// iii. the store resolver resolving store IDs to store descriptors;
		// iv.  view over what span configs apply to what keyspans;
		//
		// TODO(irfansharif): For (iii) and (iv) we might not have a
		// point-in-time snapshot of the data.
		// - For (iii) it's possible that as we iterate through the set of range
		//   descriptors, a few of which refer to some store S, we're racing
		//   against that newly-added store's info not yet being available
		//   through gossip. This is exceedingly unlikely, but if we see it
		//   happen, we can expose some snapshot of the StoreResolver state like
		//   we have for liveness.
		// - For (iv) too we're not grabbing a read lock over the backing
		//   spanconfig.KVSubscriber while reading off each span config, so it's
		//   possible we generate the report for two range descriptors with span
		//   configs from different points in time. If this too becomes a
		//   problem, we can explicitly generate a snapshot like we do for
		//   liveness.
```
</blockquote></details>

Thanks! It's worth spelling out the guarantees we make (and the ones we don't) on the interface itself -- specifically, it's worth calling out we're not using a point-in-time view of the span config state. 
___
*[`pkg/spanconfig/spanconfigreporter/reporter.go` line 111 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NFA_Iak67RIK6I8n-ar:-NHjaIhx0f9ADlTE6y1G:bszyg9z) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c1de07e122ac8284febed7d183494a56219cdd4/pkg/spanconfig/spanconfigreporter/reporter.go#L111)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Message got clipped off from earlier, but meant to say "unavailable node IDs so callers can use library functions (not written) to determine exactly why the the specific range is unavailable/non-conformant". Done.
</blockquote></details>

Yeah, I figured the gist. 
___
*[`pkg/spanconfig/spanconfigreporter/reporter.go` line 115 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NHXGCmF6oo7KsizWOx6:-NHj_7W61pd2w6AsegIb:bwuv9up) ([raw file](https://github.com/cockroachdb/cockroach/blob/ddb99140906c6f394a1b249a9d527dde64890f4d/pkg/spanconfig/spanconfigreporter/reporter.go#L115)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Great catch! Indeed it can, and that's not something we want. Changed interfaces around slightly to explicitly capture a snapshot of liveness state so such flip-flopping is not possible.
</blockquote></details>

Nice! 
___
*[`pkg/spanconfig/spanconfigreporter/reporter.go` line 121 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NHjZoXN1rK5fpv2sRGV:-NHjZoXN1rK5fpv2sRGW:b-wzpnuw) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1adc631c82f024fd59a22debb490f61a2bc0eda/pkg/spanconfig/spanconfigreporter/reporter.go#L121)):*
> ```Go
> ) (roachpb.SpanConfigConformanceReport, error) {
> 	report := roachpb.SpanConfigConformanceReport{}
> 	unavailableNodes := make(map[roachpb.NodeID]struct{})
> ```

Do we still need this now that we have the `isLiveMap`? Could we instead use that when constructing the response? 


<!-- Sent from Reviewable.io -->

irfansharif reviewed (commented):
TFTR!

bors r+

> Dismissed @arulajmani from a discussion.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-:-NHjfuWxDmBu2VYeMTZ6:b-oduykh)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @arulajmani)

___
*[`pkg/spanconfig/spanconfig.go` line 356 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NFAnpKS5pPFIU4zR1TC:-NHjhe-eDc0hQvqQsxgW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c1de07e122ac8284febed7d183494a56219cdd4/pkg/spanconfig/spanconfig.go#L356)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Thanks! It's worth spelling out the guarantees we make (and the ones we don't) on the interface itself -- specifically, it's worth calling out we're not using a point-in-time view of the span config state. 
</blockquote></details>

Done.
___
*[`pkg/spanconfig/spanconfigreporter/reporter.go` line 121 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/90016#-NHjZoXN1rK5fpv2sRGV:-NHjhs2C0PXnWIPsx0rp:b-q96705) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1adc631c82f024fd59a22debb490f61a2bc0eda/pkg/spanconfig/spanconfigreporter/reporter.go#L121)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Do we still need this now that we have the `isLiveMap`? Could we instead use that when constructing the response? 
</blockquote></details>

Not strictly, but I'll still keep things as is. Right now the list of unavailable node IDs are only those that definitely hold replicas of the keys spans requested, instead of the full set of all unavailable nodes.


<!-- Sent from Reviewable.io -->

irfansharif commented:
:(
